### PR TITLE
Bernoulli Velocity-Pressure Consistency: cross-output physics constraint

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Add a soft Bernoulli equation constraint that enforces consistency between the predicted velocity (Ux, Uy) and pressure (p) fields on surface nodes. The model currently predicts these three quantities independently — there is no explicit loss connecting them. The Bernoulli equation states: p + 0.5*rho*(Ux^2 + Uy^2) = p_total = const along streamlines. A soft constraint penalizing violations of this relationship on surface nodes provides cross-output physics information that flows into the backbone via gradients.

**Why this should work:** Physics constraints are our proven winning direction (Kutta TE: p_oodc -2.9%). The Bernoulli constraint is the ONLY multi-output consistency signal — it connects velocity and pressure predictions, forcing them to be physically coherent. Currently, PCGrad only handles gradient conflicts between dataset splits, not between output channels. This is a fundamentally different kind of constraint.

**Note:** Bernoulli is strictly inviscid. The viscous correction is O(Re^{-1/2}) in the boundary layer (~5% for Re=500K), so Bernoulli as a SOFT constraint is appropriate — it provides signal without over-constraining.

### Implementation

Add `--bernoulli_constraint` and `--bernoulli_weight` flags.

**Step 1: Compute the Bernoulli residual on surface nodes.**

After the model predicts Ux, Uy, and p at surface nodes (in physical space, AFTER inverse asinh transform):

```python
# p_pred: predicted pressure at surface nodes [N_surf]
# ux_pred, uy_pred: predicted velocity at surface nodes [N_surf]
# Umag: freestream velocity magnitude (from batch data)

# Dynamic pressure from predicted velocity
q_pred = 0.5 * (ux_pred**2 + uy_pred**2)

# Total pressure (should be approximately constant = 0.5 * Umag^2 for inviscid)
p_total_pred = p_pred + q_pred
p_total_ref = 0.5 * Umag**2  # reference total pressure

# Bernoulli residual: how far from constant total pressure
L_bernoulli = torch.mean(torch.abs(p_total_pred - p_total_ref))
```

**Step 2: Normalize the residual.**

The raw residual has units of pressure. Normalize by the reference dynamic pressure to make the loss scale-invariant:

```python
L_bernoulli_normalized = L_bernoulli / (0.5 * Umag**2 + 1e-8)
```

**Step 3: Add to total loss.**

```python
total_loss = main_loss + bernoulli_weight * L_bernoulli_normalized
```

**IMPORTANT:** Apply ONLY to surface nodes, NOT volume nodes. Volume nodes in the boundary layer have viscous effects that break Bernoulli. Surface nodes at the boundary are approximately inviscid outside the thin BL.

**IMPORTANT:** The asinh transform must be inverted before computing Bernoulli. The model predicts in asinh space. You need to convert to physical pressure/velocity first: `p_physical = sinh(p_asinh * asinh_scale)` (check the existing inverse transform in train.py).

### Training Setup

```bash
cd cfd_tandemfoil && python train.py \
  --agent fern --wandb_name "fern/bernoulli-constraint-w01-s42" \
  --wandb_group bernoulli-constraint --seed 42 \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --bernoulli_constraint --bernoulli_weight 0.1
```

Run both seeds (42, 73) with weight=0.1.

### What to Watch

- **All 4 metrics** — this constraint affects the ENTIRE prediction pipeline
- If training becomes unstable (loss spikes), reduce weight to 0.01
- Monitor the Bernoulli residual itself (log it to W&B) — it should decrease over training
- The constraint is strongest in regions where the flow is nearly inviscid (away from stagnation, TE). If metrics improve everywhere EXCEPT near TE/stagnation, the Bernoulli constraint is working as expected

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.872** | < 11.872 |
| p_oodc | 7.459 | < 7.459 |
| **p_tan** | **26.319** | < 26.319 |
| **p_re** | **6.229** | < 6.229 |

Baseline PR: #2357. W&B: aycq1m8m (s42), 9sk276v6 (s73).

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Bernoulli Velocity-Pressure Consistency Constraint (weight=0.1, 2 seeds)

| Metric | Baseline (2-seed avg) | Ours (2-seed avg) | Delta | Seed 42 | Seed 73 |
|--------|----------------------|-------------------|-------|---------|---------|
| **p_in** | **11.872** | **15.900** | **+4.028** | 19.0 | 12.8 |
| p_oodc | 7.459 | 10.350 | +2.891 | 12.6 | 8.1 |
| **p_tan** | **26.319** | **28.800** | **+2.481** | 30.5 | 27.1 |
| p_re | 6.229 | 8.250 | +2.021 | 9.7 | 6.8 |

**Note:** Seed 42 experienced training instability (val loss spikes ~epoch 63-65), leading to a much worse best checkpoint. Seed 73 trained stably and is closer to baseline, but still regressed across all metrics.

### Training Details
- **Command**: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 --bernoulli_constraint --bernoulli_weight 0.1 --wandb_group bernoulli-constraint --seed {42,73}`
- **W&B runs**: `jmi0w4eb` (seed 42), `76jyesmh` (seed 73)
- **Peak VRAM**: ~40.5 GB (no significant overhead)
- **Epoch time**: ~70s (slightly faster than baseline)
- **Best epoch**: 155 (both seeds)

### What Happened

**The constraint significantly hurt performance.** Even seed 73 (which trained stably) regressed across all metrics. Seed 42 was additionally impacted by training instability with visible val loss spikes.

**Analysis:**

1. **Bernoulli residual early in training is very large.** At epochs 1-10, the model predicts poor Ux, Uy, p. The Bernoulli residual = |p + 0.5(Ux²+Uy²) - 0.5*Umag²| / q_ref can be >>1 when the model hasn't converged yet. This creates a large loss signal that competes with the main reconstruction loss, potentially disrupting the carefully tuned warmup dynamics.

2. **The constraint is mathematically valid only on streamlines, not at arbitrary surface nodes.** On the no-slip surface, Ux=Uy=0 by definition, so `0.5*(Ux²+Uy²) = 0`. The Bernoulli equation then reduces to `p = p_total = 0.5*Umag²` — which is the stagnation pressure, true only AT the stagnation point, not everywhere on the surface. On most of the surface, the boundary condition is `p + 0 ≈ p_stag + 0` which just penalizes deviation of pressure from stagnation pressure. This is incorrect physics.

3. **The intended constraint is on the outer edge of the boundary layer, not on the wall.** Applying Bernoulli to surface nodes is physically wrong — the no-slip condition means velocity is zero at the wall. The correct application would be to field nodes just outside the BL, but we don't have a clean mask for those.

4. **Seed 42 instability:** The Bernoulli loss injects high-variance gradients through the denormalization chain (multiply by Umag^2), which scales gradients by the squared freestream velocity. This can cause gradient spikes, especially for high-Re samples.

### Bug Note

The final verification (pipeline vs manual denorm check) shows NaN for both seeds — this appears to be a pre-existing issue with the verification code when `cfg.surface_refine=True` and checkpoint loading, not related to the Bernoulli changes.

### Suggested Follow-ups

1. **Apply to field nodes only** (vol_mask), not surface nodes. Outside the BL, Bernoulli holds. This requires a distance-from-surface threshold (e.g., dist > 0.01) to exclude the BL.
2. **Ramp-in late in training** (e.g., only after epoch 50) once the model has converged enough for Bernoulli to be meaningful.
3. **Use a much smaller weight** (0.001-0.01) to prevent destabilization.
4. **Clip the Bernoulli gradient** to prevent spike propagation.
5. **Consider pressure-velocity coupling via a different route**: instead of Bernoulli, use `∇p ≈ -ρ(U·∇U)` (the Euler momentum equation) as a local constraint on gradients.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2387 because ADVISOR: Closing — clear dead end. All metrics significantly regressed (p_in +34%, p_oodc +39%, p_re +32%). Fern's analysis is excellent: applying Bernoulli at surface nodes is physically wrong because no-slip (Ux=Uy=0) makes it degenerate to p=p_stagnation everywhere. The constraint is only valid outside the BL, which we can't cleanly mask. Training instability from high-variance Bernoulli gradients confirms this approach is fundamentally flawed. Fern is already working on #2397 (Stochastic Cross-Foil Attention) — good experiment, learned something useful about physics constraints on walls..
